File: app/src/main/java/io/github/summerdez/asmrplayer/data/download/DlsiteDownloadNotifications.java
package io.github.summerdez.asmrplayer.data.download;

import io.github.summerdez.asmrplayer.R;
import io.github.summerdez.asmrplayer.data.*;
import io.github.summerdez.asmrplayer.data.remote.*;
import io.github.summerdez.asmrplayer.data.download.*;
import io.github.summerdez.asmrplayer.data.files.*;
import io.github.summerdez.asmrplayer.domain.*;
import io.github.summerdez.asmrplayer.domain.model.*;
import io.github.summerdez.asmrplayer.playback.*;
import io.github.summerdez.asmrplayer.presentation.*;
import io.github.summerdez.asmrplayer.ui.*;
import io.github.summerdez.asmrplayer.ui.activity.*;
import io.github.summerdez.asmrplayer.ui.components.*;
import io.github.summerdez.asmrplayer.ui.screens.*;
import io.github.summerdez.asmrplayer.ui.theme.*;
import io.github.summerdez.asmrplayer.ui.util.*;
import io.github.summerdez.asmrplayer.di.*;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;

final class DlsiteDownloadNotifications {
    static final int NOTIFICATION_ID = 2001;

    private static final String CHANNEL_ID = "dlsite_download";

    private DlsiteDownloadNotifications() {
    }

    static void ensureChannel(Context context) {
        NotificationManager manager = context.getSystemService(NotificationManager.class);
        if (manager == null) {
            return;
        }
        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "DLsite 下载",
                NotificationManager.IMPORTANCE_LOW);
        channel.setDescription("DLsite 作品下载与导入");
        manager.createNotificationChannel(channel);
    }

    static Notification build(Context context, String title, String status) {
        return build(context, title, status, 0, 0);
    }

    static Notification buildSummary(Context context, DlsiteDownloadState state) {
        DlsiteDownloadSummary summary = state == null ? new DlsiteDownloadSummary() : state.getSummary();
        String title = summary.getVisible()
                ? "DLsite 下载 · 正在下载 " + summary.getRunningCount() + " / " + summary.getTotalCount() + " 项"
                : "DLsite 下载";
        String text = summary.getProgressPercent() == null
                ? formatBytes(summary.getSpeedBytesPerSecond()) + "/s"
                : summary.getProgressPercent() + "% · " + formatBytes(summary.getSpeedBytesPerSecond()) + "/s · "
                + formatBytes(summary.getBytesDownloaded()) + " / " + formatBytes(summary.getTotalBytes());
        return build(context, title, text, summary.getProgressPercent() == null ? 0 : summary.getProgressPercent(), summary.getTotalBytes());
    }

    static Notification build(Context context, String title, String status, int progressPercent, long totalBytes) {
        Intent openIntent = new Intent(context, MainActivity.class);
        PendingIntent contentIntent = PendingIntent.getActivity(
                context,
                0,
                openIntent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        Notification.Builder builder = new Notification.Builder(context, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(TextUtils.isEmpty(title) ? "DLsite 下载" : title)
                .setContentText(TextUtils.isEmpty(status) ? "下载中" : status)
                .setContentIntent(contentIntent)
                .setOngoing(true)
                .setOnlyAlertOnce(true);
        if (totalBytes > 0L) {
            builder.setProgress(100, progressPercent, false);
        }
        return builder.build();
    }

    static void update(Context context, String title, String status) {
        NotificationManager manager = context.getSystemService(NotificationManager.class);
        if (manager == null) {
            return;
        }
        manager.notify(NOTIFICATION_ID, build(context, title, status));
    }

    static void updateSummary(Context context, DlsiteDownloadState state) {
        NotificationManager manager = context.getSystemService(NotificationManager.class);
        if (manager == null) {
            return;
        }
        manager.notify(NOTIFICATION_ID, buildSummary(context, state));
    }

    private static String formatBytes(long bytes) {
        if (bytes <= 0L) {
            return "--";
        }
        final String[] units = {"B", "KB", "MB", "GB"};
        double value = bytes;
        int unit = 0;
        while (value >= 1024.0 && unit < units.length - 1) {
            value /= 1024.0;
            unit++;
        }
        if (unit == 0) {
            return ((long) value) + " " + units[unit];
        }
        return String.format(java.util.Locale.US, "%.1f %s", value, units[unit]);
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/data/download/DlsiteDownloadNotifications.kt
package io.github.summerdez.asmrplayer.data.download

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.text.TextUtils
import io.github.summerdez.asmrplayer.R
import io.github.summerdez.asmrplayer.data.DlsiteDownloadState
import io.github.summerdez.asmrplayer.data.DlsiteDownloadSummary
import io.github.summerdez.asmrplayer.ui.activity.MainActivity
import java.util.Locale

object DlsiteDownloadNotifications {
    const val NOTIFICATION_ID: Int = 2001

    private const val CHANNEL_ID = "dlsite_download"

    @JvmStatic
    fun ensureChannel(context: Context?) {
        val manager = context?.getSystemService(NotificationManager::class.java) ?: return
        val channel = NotificationChannel(
            CHANNEL_ID,
            "DLsite 下载",
            NotificationManager.IMPORTANCE_LOW,
        )
        channel.description = "DLsite 作品下载与导入"
        manager.createNotificationChannel(channel)
    }

    @JvmStatic
    fun build(context: Context?, title: String?, status: String?): Notification {
        return build(context, title, status, 0, 0L)
    }

    @JvmStatic
    fun buildSummary(context: Context?, state: DlsiteDownloadState?): Notification {
        val summary = state?.summary ?: DlsiteDownloadSummary()
        val title = if (summary.visible) {
            "DLsite 下载 · 正在下载 " + summary.runningCount + " / " + summary.totalCount + " 项"
        } else {
            "DLsite 下载"
        }
        val progressPercent = summary.progressPercent
        val text = if (progressPercent == null) {
            formatBytes(summary.speedBytesPerSecond) + "/s"
        } else {
            progressPercent.toString() + "% · " +
                formatBytes(summary.speedBytesPerSecond) + "/s · " +
                formatBytes(summary.bytesDownloaded) + " / " +
                formatBytes(summary.totalBytes)
        }
        return build(context, title, text, progressPercent ?: 0, summary.totalBytes)
    }

    @JvmStatic
    fun build(
        context: Context?,
        title: String?,
        status: String?,
        progressPercent: Int,
        totalBytes: Long,
    ): Notification {
        val openIntent = Intent(context, MainActivity::class.java)
        val contentIntent = PendingIntent.getActivity(
            context,
            0,
            openIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )

        val builder = Notification.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentTitle(if (TextUtils.isEmpty(title)) "DLsite 下载" else title)
            .setContentText(if (TextUtils.isEmpty(status)) "下载中" else status)
            .setContentIntent(contentIntent)
            .setOngoing(true)
            .setOnlyAlertOnce(true)
        if (totalBytes > 0L) {
            builder.setProgress(100, progressPercent, false)
        }
        return builder.build()
    }

    @JvmStatic
    fun update(context: Context?, title: String?, status: String?) {
        val manager = context?.getSystemService(NotificationManager::class.java) ?: return
        manager.notify(NOTIFICATION_ID, build(context, title, status))
    }

    @JvmStatic
    fun updateSummary(context: Context?, state: DlsiteDownloadState?) {
        val manager = context?.getSystemService(NotificationManager::class.java) ?: return
        manager.notify(NOTIFICATION_ID, buildSummary(context, state))
    }

    private fun formatBytes(bytes: Long): String {
        if (bytes <= 0L) {
            return "--"
        }
        val units = arrayOf("B", "KB", "MB", "GB")
        var value = bytes.toDouble()
        var unit = 0
        while (value >= 1024.0 && unit < units.size - 1) {
            value /= 1024.0
            unit++
        }
        return if (unit == 0) {
            value.toLong().toString() + " " + units[unit]
        } else {
            String.format(Locale.US, "%.1f %s", value, units[unit])
        }
    }
}
